### PR TITLE
8215537: [TEST_BUG] java/awt/print/PrinterJob/LandscapeStackOverflow.java fails by timeout

### PR DESCRIPTION
**Pull request by prsadhuk**
Test fails with timeout and it seems likely due to non-availability of printer in the test system. We should just update the test to check for default (or available) printers and return if there is no printer.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed

### Issue
 * [JDK-8215537](https://bugs.openjdk.java.net/browse/JDK-8215537): [TEST_BUG] java/awt/print/PrinterJob/LandscapeStackOverflow.java fails by timeout

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1927/head:pull/1927`
`$ git checkout pull/1927`

### CODE REVIEW

**Comment by mrserb** [comment]
If the printer is not found then why the "PrinterException: No print service found" is not thrown?

**Comment by prsadhuk** [comment]
IIUC, In windows there can be non-real printers like "Send To OneNote", "Fax" which falls in Printers/Scanner list but not exactly printer so PrinterException is not thrown.

**Comment by mrserb** [comment]
> IIUC, In windows there can be non-real printers like "Send To OneNote", "Fax" which falls in Printers/Scanner list but not exactly printer so PrinterException is not thrown.

I am not sure that the fax is related here, but the "printjob.print()" should throw such an exception if printjob.getPrintService() is null (so we do not have any useful printers). Do you have a chance to reproduce this bug?

BTW the suggestion in the JBS was to fail the test if no printers are found.

File: test/jdk/java/awt/print/PrinterJob/LandscapeStackOverflow.java [hunk 1]
```diff
@@ -43,6 +45,10 @@
 
     public static final void main( String[] parameters ) {
         PrinterJob printjob = PrinterJob.getPrinterJob();
+        PrintService defaultPrtSrv = PrintServiceLookup.lookupDefaultPrintService();
+        if (printjob.getPrintService() == null || defaultPrtSrv == null) {
+            throw new RuntimeException("No printer found");
```
Comment (prrace) [inline test/jdk/java/awt/print/PrinterJob/LandscapeStackOverflow.java:50]: We need to make this more robust.
The checks should be before we create the pritner job and should be to find ANY printer and none
throws the excepton, but if there is only a non-default one, set that as the printer for the job after creating the job.

That may not matter on Windows but on other platforms I am sure you can have printers, but not a default one.

Comment (prsadhuk) [reply test/jdk/java/awt/print/PrinterJob/LandscapeStackOverflow.java]: I was of the understanding that if there are multiple, then we have to choose one as default so defaultPrintService will not be null. But, the above code looks more apt.

File: test/jdk/java/awt/print/PrinterJob/LandscapeStackOverflow.java [hunk 2]
```diff
@@ -35,14 +35,22 @@
...
+import javax.print.PrintService;
 import javax.print.attribute.HashPrintRequestAttributeSet;
 import javax.print.attribute.PrintRequestAttributeSet;
 import javax.print.attribute.standard.OrientationRequested;
 
 public class LandscapeStackOverflow {
 
-    public static final void main( String[] parameters ) {
+    public static final void main( String[] parameters ) throws Exception {
+        PrintService[] printService = PrinterJob.lookupPrintServices();
```
Comment (mrserb) [inline test/jdk/java/awt/print/PrinterJob/LandscapeStackOverflow.java:46]: As far as I understand and comment in my previouse comment, if the lookupPrintServices() return empty array then the "PrinterJob.print()" will throw the same PrinterException("No print service found");

And if default printer is not set, and we have some other printers then "PrinterJob.print()" will select some non-default printer already, isn't it?

I meant that we just duplicate an existed checks, and both of them wll not work. If the system has some non-real printers like XPS, or "print to PDF" then such printers will be count  by the "PrinterJob.lookupPrintServices()" so no exception will be thrown and the test just hands suggestion the user to save the printed document somewhere. The hang occur exactly where it is reported in JBS "WPrinterJob._startDoc".

I can reproduce it on the freshly installed Windows.

Comment (prsadhuk) [reply test/jdk/java/awt/print/PrinterJob/LandscapeStackOverflow.java:46]: I'm not sure we can avoid this if user selects "print to pdf" as default printservice or if it is the only printservice available. Maybe converting it to manual is the way to go.

Comment (mrserb) [reply test/jdk/java/awt/print/PrinterJob/LandscapeStackOverflow.java:46]: We can update the documentation "testing.html" about the "printer" keyword and describe how to skip/run the tests marked by this keyword. or we can just close the bug as not-a-defect.

**Comment by mrserb** [comment]
Not sure that one more manual test is needed, the run of the test marked by the printer keyword already required to have a configured printer. Otherwise, why do we need that keyword at all?

**Comment by prsadhuk** [comment]
That printer keyword seems to be just a hint/suggestion as it does not force user to do anything, so we can still get this kind of problem. Making it manual at least will let user choose which printer he actually wants to work with.

**Comment by mrserb** [comment]
That's not a hint, it is similar to the headful keyword:
 - "headful" requires the proper configured system and accessible Desktop
 - "printer" requires the properly configured printer
https://github.com/openjdk/jdk/blob/0957d9eb1452b953c1a700ce7ea41155c22bc00e/test/jdk/TEST.ROOT#L14

**Comment by prsadhuk** [comment]
OK. But it's a keyword for jtreg, not useful for standalone execution as you have seen that one can get the problem running on freshly installed windows system. Making it manual will overcome standalone execution problem too.

**Comment by mrserb** [comment]
It is similar to run the headful tests on windows/macOS w/o system preparation or run the tests on Linux w/o Xserver. Or similar to running the tests when the system shortcuts are enabled, will you change all tests which press combinations of keys to the manual tests? I assume no since these are the requirements for the tests run.

**Comment by prsadhuk** [comment]
Ok. I will close this as not-an-issue then.

**Pull request closed without merging**